File: solutions/1147-flip-columns-for-maximum-number-of-equal-rows/flip-columns-for-maximum-number-of-equal-rows.java
class Solution {
    public int maxEqualRowsAfterFlips(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        Map<String, Integer> patterns = new HashMap<>();
        
        StringBuilder keyBuilder = new StringBuilder(cols);
        
        for (int[] row : matrix) {
            keyBuilder.setLength(0);
            int flip = row[0];
            
            int j = 0;
            for (; j + 3 < cols; j += 4) {
                int val1 = row[j] ^ flip;
                int val2 = row[j + 1] ^ flip;
                int val3 = row[j + 2] ^ flip;
                int val4 = row[j + 3] ^ flip;
                
                keyBuilder.append(val1).append(val2).append(val3).append(val4);
            }
            
            for (; j < cols; j++) {
                keyBuilder.append(row[j] ^ flip);
            }
            
            String key = keyBuilder.toString();
            patterns.merge(key, 1, Integer::sum);
        }
        
        return patterns.values().stream().mapToInt(Integer::intValue).max().orElse(0);
    }
}

File: solutions/1147-flip-columns-for-maximum-number-of-equal-rows/flip-columns-for-maximum-number-of-equal-rows.rs
use std::collections::HashMap;

impl Solution {
    pub fn max_equal_rows_after_flips(matrix: Vec<Vec<i32>>) -> i32 {
        let mut count: HashMap<String, i32> = HashMap::new();
        
        for row in matrix.iter() {
            let key: String = row.iter()
                .map(|&n| if row[0] == 1 { 1 - n } else { n })
                .map(|n| n.to_string())
                .collect();
                
            *count.entry(key).or_insert(0) += 1;
        }
        
        *count.values().max().unwrap()
    }
}